src/http/common.cc

```diff
@@ -83,6 +83,14 @@ namespace internal {
 
 static constexpr size_t default_body_sink_buffer_size = 32000;
 
+// Data sinks below are running "on top" of socket output stream and provide
+// reliable and handy way of generating request bodies according to selected
```

Comment:
But why **request** bodies? I think these are used in writing bodies in general, both for a server writing a reply bodies, and client preparing a request body. Right?

Reply:
Mm... no, server doesn't write content-length body like this. It either puts plain sstring as body or uses chunk- encoding sink

Reply:
Maybe I'm misunderstanding something, but this was code that predates the new "HTTP client" support, and was used in the HTTP server as well.

Reply:
Yes, I got a bit confused. But still, couldn't the server too want to send a response exactly in this new way - have a content-length *and* a body writer?

You might argue that the server doesn't need this new feature - it can simply use chunked-encoding. But the client doesn't "need" it either - it too could use chunked-encoding but you said that there is code (s3) that doesn't like chunked encoding. Maybe there is such client code as well, and the server might also want an ability to return the output of body_writer without chunked encoding?

Reply:
> Yes, I got a bit confused. But still, couldn't the server too want to send a response exactly in this new way - have a content-length and a body writer?

It could, but ...

> Maybe there is such client code as well, and the server might also want an ability to return the output of body_writer without chunked encoding?

... I don't know -- is there anything the the protocol using which http client could say "I don't accept chunked-encoded replies, sorry"? If there is, then server could benefit from it too.

Reply:
There is no such thing in HTTP 1.1 - chunked encoding is part of the protocol, and neither a server nor a client can choose not to implement it (though as you noted, S3 still decided to not implement it).

But it is conceivable that sometimes sending a response with content-length has benefits. For example, I would guess that graphical browsers show nice progress percentage for a download with content-length (e.g., tell you that the download is "37% done"), and less-nice user interface when downloading with chunked encoding (just "downloaded 123 MB..." without an indication how much there is still left).
So it makes sense for a server that does know the response size in advance (e.g., because the response is read from disk, or from some remote server, but its size is known) to use it, and not use chunked encoding.